### PR TITLE
Fasta file

### PR DESCRIPTION
**Pull request by pope-capable**
Fasta Method

### CODE REVIEW

File: Analyser/src/bsa_analyser/github/io/FileValidator.java [hunk 1]
```diff
@@ -5,24 +5,79 @@
...
+
+    List<String> Fasta_Content = new ArrayList<String>();
+        if (user_file.getName().endsWith("fasta") || user_file.getName().endsWith("fa")) {
+            String line = "";
+            try {
+
+                BufferedReader br = new BufferedReader(new FileReader((user_file)));
+                while (((line = br.readLine()) != null)) {
+
+                    if ((!line.contains("A|T|G|C|N")) && (!line.startsWith(">"))) {
```
Comment (jkelly282) [inline Analyser/src/bsa_analyser/github/io/FileValidator.java]: This line here is checking if the file does not contain the string "A|T|G|C|N" not the individual bases. So this is always going to be True unless the Fasta file actually contains that string. Also a fasta file has two types of lines either it starts with '>' or has the nucleotides in the sting. This logic is checking for both.

File: Analyser/src/bsa_analyser/github/io/FileValidator.java [hunk 2]
```diff
@@ -5,24 +5,79 @@
...
+                    }
+                    Scanner inFile1 = new Scanner(new FileReader(user_file));//.useDelimiter("\\>");
+
+                    while (inFile1.hasNext()) {
+                        String token1 = inFile1.nextLine();
+                        Fasta_Content.add(token1);
+                    }
+
+
+                    System.out.println(Fasta_Content);
```
Comment (jkelly282) [inline Analyser/src/bsa_analyser/github/io/FileValidator.java]: Why is this method printing out the whole fasta to the command line ?

File: Analyser/src/bsa_analyser/github/io/FileValidator.java [hunk 3]
```diff
@@ -5,24 +5,79 @@
...
+                                        + " (error code 1)");
+
+                                if (user_option == 0) {
+                                    BSA_Visualisation.upload_files();
+                                    b = true;
+                                    break;
+                                } else {
+                                    JOptionPane.showMessageDialog(null, "This Fasta has not being verified");
+                                }
+                                System.out.println("WARNING - File not verified as being .vcf (error code 2)");
```
Comment (jkelly282) [inline Analyser/src/bsa_analyser/github/io/FileValidator.java]: This should be .fasta/.fa

File: Analyser/src/bsa_analyser/github/io/FileValidator.java (hunk 1 above)
Comment (jkelly282) [inline Analyser/src/bsa_analyser/github/io/FileValidator.java]: Is this line correct seems to be letting wrong fasta files through !

File: Analyser/src/bsa_analyser/github/io/FileValidator.java [hunk 4]
```diff
@@ -5,24 +5,79 @@
...
+
+                    if ((!line.contains("A|T|G|C|N")) && (!line.startsWith(">"))) {
+                        //while (((line = br.readLine() != null)) {
+                        for (int i = 0; i < line.length(); i++) {
+                            if (!(line.substring(i, i + 1).matches(("|A|T|G|C|N|a|t|g|c|n")))) {
+                                int user_option = JOptionPane.showConfirmDialog(null, "This Fasta file is corrupt"
+                                        + " This means the content may not be a Fasta file format. Do you wish to upload another file?"
+                                        + " (error code 1)");
+
+                                if (user_option == 0) {
```
Comment (jkelly282) [inline Analyser/src/bsa_analyser/github/io/FileValidator.java]: Having this logic in  a while loop means that if the user doesn't want to upload the file it keeps asking if they want to upload another file - it is impossible to break out of the loop!

**Review by jkelly282 — CHANGES_REQUESTED** [review]
We have tested it with a mock fasta file that only contains 'ACG' and it seems to work - would you be able to fix this so if there is a file like this it does not allow the upload. Please can you also leave comments on the code - we found it quite hard to read the code ! 

Also take the file chooser out of the while loop - otherwise it leads to an endless loop !

**Comment by pope-capable** [comment]
Okay will, resolve this, right away

**Pull request closed without merging**